fundamentals-section/fundamentals.md

```diff
@@ -0,0 +1,300 @@
+- Function composition is easy to do in go. 
+- Deferred functions are useful for guaranteed clean up tasks. 
+
+[Playground](https://go.dev/play/p/fDUKmkVvezv)
+
+```go
+func sayHi(hosts []string) {
+	greeting := fmt.Sprintf("Hello, friends! We're %s and %s.", hosts[0], hosts[1])
+	fmt.Println(greeting)
+
+}
+
+func main() {
+	defer func() {
+		fmt.Println("Goodbye, friends!")
+	}()
+	hosts := []string{"Adelina", "Liam"}
+	sayHi(hosts)
+}
+
+```
+
+## Custom types
+- Go is not an object oriented language, as it does not support type hierarchy. 
+- Structs allow us to build custom types and behaviours.
+- They are a collection of fields, which can be partially initialised.
+- Custom types can also methods with a special receiver argument that is the implicit first argument of the method.
+
+[Playground](https://go.dev/play/p/DDT86lbYhhz)
+```go
+type host struct {
+	name string
+}
+
+func (h host) sayHi() string {
+	return fmt.Sprintf("Hello, friends! I'm %s.", h.name)
+}
+
+func main() {
+	hosts := []host{
+		host{name: "Adelina"},
+		host{name: "Liam"},
+	}
+	for _, h := range hosts {
+		fmt.Println(h.sayHi())
+	}
+
+}
+```
+
+## Visibility
+- Go code is organised in packages, which control the visibility of the variables, types and functions they contain.
+- A folder may only contain a single package, but the package does not need to be named after the directory .
+- Names can only be used once inside the same package.
+- Runnable programs have a main function defined in a main package.
+- We can export fields outside their package by capitalising the first letter of their name. 
+
+[Playground](https://go.dev/play/p/m0U35Q6fWMJ)
+```go
+type host struct {
+	name string
+}
+
+func (h host) sayHi() string {
+	return fmt.Sprintf("Hello, friends! I'm %s.", h.name)
+}
+
+func (h host) name() {
+	return h.name
+}
+```
+
+## Interfaces
+- Interfaces are collections of method signatures. 
+- They are automatically implemented by the compiler on types which satisfy the entire collection of methods. 
+- They are the primary way of implementing polymorphism in Go.
+- Interfaces are often exported, while the structs remain visible only inside the package. 
+
+[Playground](https://go.dev/play/p/GHiwVPPVdPF)
+
+```go
+type Host interface {
+	SayHi() string
+}
+
+type experiencedHost struct {
+	name string
+}
+
+func (eh experiencedHost) SayHi() string {
+	return fmt.Sprintf("Hello again, friends! I'm %s.", eh.name)
+}
+
+type newHost struct {
+	name string
+}
+
+func (nh newHost) SayHi() string {
+	return fmt.Sprintf("Hello, friends! I'm %s and I'm new here.", nh.name)
+}
+
+func main() {
+	hosts := []Host{
+		newHost{name: "Adelina"},
+		experiencedHost{name: "Liam"},
+	}
+	for _, h := range hosts {
+		fmt.Println(h.SayHi())
+	}
+}
+```
+
+## Goroutines
+- Goroutines are known as lightweight threads. They are used to functions concurrently inside our Go programs.
+- We instruct the Go runtime to run a function in a new goroutine by using the go keyword.
```

Comment:
Markdown notation on "go" for descriptive reading. Use backticks to encase it like: `go`